concordium-std/src/test_infrastructure.rs

```diff
@@ -530,3 +533,222 @@ pub fn report_error(message: &str, filename: &str, line: u32, column: u32) {
+}
+
+impl<T: convert::AsMut<[u8]>> Write for ContractStateTest<T> {
+    type Err = ContractStateError;
+
+    fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Err> {
+        // Write should only write as much as there is space currently, and not
+        // automatically extend the data.
+        // In Vec<u8> based implementations the latter behaviour would be reasonable,
+        // however that is not how the chain implementation behaves, and thus this one
+        // should not behave like that either.
+        let data = &mut self.cursor.data.as_mut()[self.cursor.offset..];
+        let to_write = cmp::min(data.len(), buf.len());
+        data.copy_from_slice(&buf[..to_write]);
+        self.cursor.offset += to_write;
+        Ok(to_write)
+    }
+}
+
+impl<T: AsMut<Vec<u8>> + AsMut<[u8]> + AsRef<[u8]>> HasContractState<ContractStateError>
+    for ContractStateTest<T>
+{
+    type ContractStateData = T;
+
+    fn open(data: Self::ContractStateData) -> Self {
+        Self {
+            cursor: Cursor::new(data),
+        }
+    }
+
+    fn size(&self) -> u32 { self.cursor.data.as_ref().len() as u32 }
+
+    fn truncate(&mut self, new_size: u32) {
+        if self.size() > new_size {
+            let new_size = new_size as usize;
+            let data: &mut Vec<u8> = self.cursor.data.as_mut();
+            data.truncate(new_size);
+            if self.cursor.offset > new_size {
+                self.cursor.offset = new_size
+            }
+        }
+    }
+
+    fn reserve(&mut self, len: u32) -> bool {
+        if len <= constants::MAX_CONTRACT_STATE_SIZE {
+            if self.size() < len {
+                let data: &mut Vec<u8> = self.cursor.data.as_mut();
+                data.resize(len as usize, 0u8);
+            }
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl Default for ContractStateError {
+    fn default() -> Self { Self::Default }
+}
+
+impl From<num::TryFromIntError> for ContractStateError {
+    fn from(_: num::TryFromIntError) -> Self { ContractStateError::Overflow }
+}
+
+impl<T: AsRef<[u8]>> Seek for ContractStateTest<T> {
+    type Err = ContractStateError;
+
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64, Self::Err> {
+        use ContractStateError::*;
+        match pos {
+            SeekFrom::Start(x) => {
+                // We can set the position to just after the end of the current length.
+                let new_offset = x.try_into()?;
+                if new_offset <= self.cursor.data.as_ref().len() {
+                    self.cursor.offset = new_offset;
+                    Ok(x)
+                } else {
+                    Err(Offset)
+                }
+            }
+            SeekFrom::End(x) => {
+                // cannot seek beyond end, nor before beginning
+                if x <= 0 {
+                    let end: u32 = self.cursor.data.as_ref().len().try_into()?;
+                    let minus_x = x.checked_abs().ok_or(Overflow)?;
```

Comment:
When reading the documentation for [`std::io::Seek`](https://doc.rust-lang.org/std/io/trait.Seek.html), it seems to be *allowed* to seek beyond the end.
Not sure how 'allowed' should be interpreted, whether it means: Results in Ok.

Reply:
The documentation for [SeekFrom](https://doc.rust-lang.org/std/io/enum.SeekFrom.html#variant.End) adds a bit more to this.

Reply:
I know, but this is our Seek trait, and our implementation. It makes it cleaner to not allow this because then we have the invariant that offset is a valid place in the state.

Reply:
And also, it states `A seek beyond the end of a stream is allowed, but behavior is defined by the implementation.`

So we can choose to say "Err" is my interpretation of it.